### PR TITLE
replaced slice with iteration

### PR DESCRIPTION
**Pull request by mightyaleksey**

### CODE REVIEW

**Comment by fanatid** [comment]
OOI, @sullenor can you paste benchmark results?

File: index.js [hunk 1]
```diff
@@ -10,7 +8,7 @@ var slice = Array.prototype.slice;
  */
 module.exports = function promisify(fn, ctx, _) {
   var argsNum = fn.length - 1;
-  var argsToPrepend = slice.call(arguments, 2);
+  var argsToPrepend = toArray(arguments).slice(2);
```
Comment (fanatid) [inline index.js]: but new code still allows leaking `arguments` :smile: 
should be:

``` js
for (var argsToPrepend = [], i = 2; i < arguments.length; ++i) argsToPrepend.push(arguments[i])
```

Comment (mightyaleksey) [inline index.js]: Do you think, that it should be inlined?

Comment (fanatid) [inline index.js]: definitely, example in bluebird after build:

``` js
function doesntLeakArguments() {
    var $_len = arguments.length;
    var args = new Array($_len); 
    for(var $_i = 0; $_i < $_len; ++$_i) {
        args[$_i] = arguments[$_i];
    }
    return args;
}
```

**Comment by mightyaleksey** [comment]
@fanatid sure,

current master:

``` bash
call-with-multiple-arguments.js
 → The sample arithmetic mean (secs): 0.001454444422953964
call-without-arguments.js
 → The sample arithmetic mean (secs): 0.0014423124365812983
```

this branch:

``` bash
arguments-iteratee.js
 → The sample arithmetic mean (secs): 7.69186052437467e-8
arguments-slice.js
 → The sample arithmetic mean (secs): 0.000004616306086168628
call-with-multiple-arguments.js
 → The sample arithmetic mean (secs): 0.0014537959604729693
call-without-arguments.js
 → The sample arithmetic mean (secs): 0.0014482428545699313
```

Actually not quite a big difference and looks more like an error.

**Comment by mightyaleksey** [comment]
> Actually not quite a big difference and looks more like an error.

Talking about async part :) The difference in the **arguments-iteratee** and **arguments-slice** is stable.

**Comment by fanatid** [comment]
wow, `toArray` looks very fast:

```
$ node -v
v5.10.0
$ cat b.js 
var benchmark = require('benchmark')

function toArray(col) {
  var args = new Array(col.length);
  for (var i = 0; i < args.length; ++i) {
    args[i] = col[i];
  }

  return args;
}

new benchmark.Suite({ onCycle: (event) => console.log(String(event.target)) })
  .add('[].slice.call(arguments)', () => {
    return (function () {
      return [].slice.call(arguments)
    })(1, 2, 3, 4, 5)
  })
  .add('toArray(arguments)', () => {
    return (function () {
      return toArray(arguments)
    })(1, 2, 3, 4, 5)
  })
  .add('for-loop', () => {
    return (function () {
      var args = new Array(arguments.length)
      for (i = 0; i < args.length; ++i) args[i] = arguments[i]
      return args
    })(1, 2, 3, 4, 5)
  })
  .add('[].slice.call(arguments, 2)', () => {
    return (function () {
      return [].slice.call(arguments, 2)
    })(1, 2, 3, 4, 5)
  })
  .add('toArray(arguments).slice(2)', () => {
    return (function () {
      return toArray(arguments).slice(2)
    })(1, 2, 3, 4, 5)
  })
  .add('for-loop with offset', () => {
    return (function () {
      var args = []
      for (i = 2; i < arguments.length; ++i) args.push(arguments[i])
      return args
    })(1, 2, 3, 4, 5)
  })
  .run()
$ node b.js 
[].slice.call(arguments) x 280,364 ops/sec ±2.28% (81 runs sampled)
toArray(arguments) x 8,887,139 ops/sec ±0.71% (85 runs sampled)
for-loop x 2,573,323 ops/sec ±0.19% (89 runs sampled)
[].slice.call(arguments, 2) x 286,014 ops/sec ±2.63% (84 runs sampled)
toArray(arguments).slice(2) x 2,858,260 ops/sec ±0.23% (91 runs sampled)
for-loop with offset x 249,243 ops/sec ±1.15% (86 runs sampled)
```

**Comment by fanatid** [comment]
maybe v8 re-compile `toArray` to "hot" function...

File: index.js (hunk 1 above)
Comment (mightyaleksey) [inline index.js]: Thanks, I've updated the version and looks like it is a bit faster 👍

**Comment by mightyaleksey** [comment]
Yeah, seems so :)

Anyways, I tried the new variant and looks like it's a bit better:

``` bash
arguments-iteratee.js
 → The sample arithmetic mean (secs): 8.127000742839326e-8
arguments-push.js
 → The sample arithmetic mean (secs): 1.7087444227219143e-7
arguments-slice.js
 → The sample arithmetic mean (secs): 0.0000062300160441625
call-with-multiple-arguments.js
 → The sample arithmetic mean (secs): 0.0013566487996386543
call-without-arguments.js
 → The sample arithmetic mean (secs): 0.0013547885377925269
```

File: index.js [hunk 2]
```diff
@@ -29,7 +29,9 @@ module.exports = function promisify(fn, ctx, _) {
   }
 
   return function promise() {
-    var args = slice.call(arguments, 0, Math.max(argsNum - argsToPrepend.length, 0));
+    for (var args = Array(Math.max(argsNum - argsToPrepend.length, 0)), i = 0; i < args.length; ++i) {
```
Comment (fanatid) [inline index.js]: `Math.max(argsNum - argsToPrepend.length, 0)` is constant, you can move this from function

Comment (mightyaleksey) [inline index.js]: Agree, thanks

**Comment by mightyaleksey** [comment]
@fanatid I like you variant of output in the benchmark example, so I decided to change mine :)
Also I decided to add extra example and delegate this job to travis. So here is the result for the current pr:

``` bash
node v0.12.13

[].slice.call(arguments)            x 1,080,505 ops/sec ±0.58% (91 runs sampled)
[].proto.slice.call(arguments)      x 1,090,071 ops/sec ±1.10% (90 runs sampled)
do-while-loop                       x 28,492,497 ops/sec ±0.88% (88 runs sampled)
for-loop with assignment            x 30,801,291 ops/sec ±0.57% (89 runs sampled)
for-loop with [].push(arguments[i]) x 13,101,769 ops/sec ±4.30% (88 runs sampled)
while-loop                          x 25,712,453 ops/sec ±0.60% (89 runs sampled)

promise call with 3 arguments       x 63,543 ops/sec ±8.04% (60 runs sampled)
promise call with zero arguments    x 38,141 ops/sec ±3.93% (71 runs sampled)
```

``` bash
node v4.3.2

[].slice.call(arguments)            x 456,249 ops/sec ±0.76% (85 runs sampled)
[].proto.slice.call(arguments)      x 468,587 ops/sec ±0.68% (91 runs sampled)
do-while-loop                       x 33,874,619 ops/sec ±0.76% (90 runs sampled)
for-loop with assignment            x 35,874,722 ops/sec ±0.53% (90 runs sampled)
for-loop with [].push(arguments[i]) x 16,948,879 ops/sec ±1.05% (88 runs sampled)
while-loop                          x 33,576,410 ops/sec ±0.55% (88 runs sampled)

promise call with 3 arguments       x 87,697 ops/sec ±6.36% (60 runs sampled)
promise call with zero arguments    x 84,800 ops/sec ±7.22% (63 runs sampled)
```

Thanks for help :)

**Comment by fanatid** [comment]
nice benchmarks :)
obviously `.push` is slower than `assignment` because memory reallocation
probably `do-while-loop` will have same numbers as `for-loop with assignment` if you change decrement to increment

**Comment by mightyaleksey** [comment]
> probably do-while-loop will have same numbers as for-loop with assignment if you change decrement to increment

I didn't know about that. You are an eye opener for me! :) I try to check it later.

Anyways, here is another result from node 5 (done by travis):

``` bash
node v5.10.1

[].slice.call(arguments)            x 433,126 ops/sec ±6.58% (83 runs sampled)
[].proto.slice.call(arguments)      x 484,534 ops/sec ±3.09% (90 runs sampled)
do-while-loop                       x 27,295,367 ops/sec ±5.74% (80 runs sampled)
for-loop with assignment            x 35,492,466 ops/sec ±1.19% (86 runs sampled)
for-loop with [].push(arguments[i]) x 17,039,609 ops/sec ±0.54% (88 runs sampled)
while-loop                          x 33,405,619 ops/sec ±0.77% (90 runs sampled)

promise call with 3 arguments       x 108,058 ops/sec ±10.97% (63 runs sampled)
promise call with zero arguments    x 101,648 ops/sec ±32.37% (66 runs sampled)
```

**Comment by mightyaleksey** [comment]
Actually, I truly believed in to `while-loop` construction. But, according to the benchmarks, it is close to the `for-loop`.

**Comment by fanatid** [comment]
> I didn't know about that. You are an eye opener for me! :) I try to check it later.

I'm not sure, even don't remember where I read about this..

**Pull request merged**